File: src/omnia/data/structure/tuple/ImmutableSeptuplet.java
package omnia.data.structure.tuple;

class ImmutableSeptuplet<T> extends ImmutableSeptuple<T, T, T, T, T, T, T> implements Septuplet<T> {

  ImmutableSeptuplet(T first, T second, T third, T fourth, T fifth, T sixth, T seventh) {
    super(first, second, third, fourth, fifth, sixth, seventh);
  }

  @Override
  public Sextuplet<T> dropFirst() {
    return Tuplet.of(second(), third(), fourth(), fifth(), sixth(), seventh());
  }

  @Override
  public Sextuplet<T> dropSecond() {
    return Tuplet.of(first(), third(), fourth(), fifth(), sixth(), seventh());
  }

  @Override
  public Sextuplet<T> dropThird() {
    return Tuplet.of(first(), second(), fourth(), fifth(), sixth(), seventh());
  }

  @Override
  public Sextuplet<T> dropFourth() {
    return Tuplet.of(first(), second(), third(), fifth(), sixth(), seventh());
  }

  @Override
  public Sextuplet<T> dropFifth() {
    return Tuplet.of(first(), second(), third(), fourth(), sixth(), seventh());
  }

  @Override
  public Sextuplet<T> dropSixth() {
    return Tuplet.of(first(), second(), third(), fourth(), fifth(), seventh());
  }

  @Override
  public Sextuplet<T> dropSeventh() {
    return Tuplet.of(first(), second(), third(), fourth(), fifth(), sixth());
  }

  @Override
  public Octuplet<T> concat(T object) {
    return Tuplet.of(first(), second(), third(), fourth(), fifth(), sixth(), seventh(), object);
  }

  @Override
  public Nonuplet<T> concat(Couple<T, T> other) {
    return Tuplet.of(first(), second(), third(), fourth(), fifth(), sixth(), seventh(), other.first(), other.second());
  }

  @Override
  public Decuplet<T> concat(Triple<T, T, T> other) {
    return Tuplet.of(first(), second(), third(), fourth(), fifth(), sixth(), seventh(), other.first(), other.second(), other.third());
  }
}

File: src/omnia/data/structure/tuple/ImmutableSeptuplet.kt
package omnia.data.structure.tuple

import java.util.function.Function
import omnia.data.structure.immutable.ImmutableList

internal class ImmutableSeptuplet<T : Any>(
  first: T,
  second: T,
  third: T,
  fourth: T,
  fifth: T,
  sixth: T,
  seventh: T
) : ImmutableSeptuple<T, T, T, T, T, T, T>(first, second, third, fourth, fifth, sixth, seventh),
  Septuplet<T> {

  override fun <R : Any> map(mapper: Function<in T, out R>): Septuplet<R> {
    return Tuplet.of(
      mapper.apply(first()),
      mapper.apply(second()),
      mapper.apply(third()),
      mapper.apply(fourth()),
      mapper.apply(fifth()),
      mapper.apply(sixth()),
      mapper.apply(seventh())
    )
  }

  override fun dropFirst(): Sextuplet<T> {
    return Tuplet.of(second(), third(), fourth(), fifth(), sixth(), seventh())
  }

  override fun dropSecond(): Sextuplet<T> {
    return Tuplet.of(first(), third(), fourth(), fifth(), sixth(), seventh())
  }

  override fun dropThird(): Sextuplet<T> {
    return Tuplet.of(first(), second(), fourth(), fifth(), sixth(), seventh())
  }

  override fun dropFourth(): Sextuplet<T> {
    return Tuplet.of(first(), second(), third(), fifth(), sixth(), seventh())
  }

  override fun dropFifth(): Sextuplet<T> {
    return Tuplet.of(first(), second(), third(), fourth(), sixth(), seventh())
  }

  override fun dropSixth(): Sextuplet<T> {
    return Tuplet.of(first(), second(), third(), fourth(), fifth(), seventh())
  }

  override fun dropSeventh(): Sextuplet<T> {
    return Tuplet.of(first(), second(), third(), fourth(), fifth(), sixth())
  }

  override fun concat(`object`: T): Octuplet<T> {
    return Tuplet.of(first(), second(), third(), fourth(), fifth(), sixth(), seventh(), `object`)
  }

  override fun concat(other: Couple<T, T>): Nonuplet<T> {
    return Tuplet.of(
      first(),
      second(),
      third(),
      fourth(),
      fifth(),
      sixth(),
      seventh(),
      other.first(),
      other.second()
    )
  }

  override fun concat(other: Triple<T, T, T>): Decuplet<T> {
    return Tuplet.of(
      first(),
      second(),
      third(),
      fourth(),
      fifth(),
      sixth(),
      seventh(),
      other.first(),
      other.second(),
      other.third()
    )
  }

  override fun iterator(): Iterator<T> {
    return toActualList().iterator()
  }

  override fun itemAt(index: Int): T {
    return toActualList().itemAt(index)
  }

  override fun indexOf(item: Any?): Int? {
    return toActualList().indexOf(item)
  }

  override fun containsUnknownTyped(item: Any?): Boolean {
    return toActualList().containsUnknownTyped(item)
  }

  private fun toActualList(): ImmutableList<T> {
    return ImmutableList.of(first(), second(), third(), fourth(), fifth(), sixth(), seventh())
  }
}